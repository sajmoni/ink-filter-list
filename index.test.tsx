import React from 'react'
import { render } from 'ink-testing-library'
import { test, expect } from 'vitest'

import timers from 'node:timers/promises'

import FilterList from './src/index.jsx'

const ARROW_UP = '\u001B[A'
const ARROW_DOWN = '\u001B[B'
const BACKSPACE = '\u007F'

const input = async (write: (data: string) => void, input: string) => {
  await timers.setTimeout(100)
  write(input)
}

test('Arrow keys and filtering', async () => {
  const { lastFrame, stdin } = render(
    <FilterList
      items={[
        {
          label: 'First',
          value: 'first',
        },
        {
          label: 'Second',
          value: 'second',
        },
        {
          label: 'Third',
          value: 'third',
        },
        {
          label: 'Fourth',
          value: 'fourth',
        },
        {
          label: 'Fifth',
          value: 'fifth',
        },
      ]}
      onSubmit={() => {}}
    />,
  )

  expect(lastFrame()).toMatchSnapshot()

  await input(stdin.write, ARROW_DOWN)

  expect(lastFrame()).toMatchSnapshot()

  await input(stdin.write, ARROW_DOWN)

  expect(lastFrame()).toMatchSnapshot()

  await input(stdin.write, ARROW_UP)

  expect(lastFrame()).toMatchSnapshot()

  await input(stdin.write, 'second')

  expect(lastFrame()).toMatchSnapshot()

  // Clear the input
  await input(stdin.write, BACKSPACE)
  await input(stdin.write, BACKSPACE)
  await input(stdin.write, BACKSPACE)
  await input(stdin.write, BACKSPACE)
  await input(stdin.write, BACKSPACE)
  await input(stdin.write, BACKSPACE)

  await input(stdin.write, 'does not exist')

  expect(lastFrame()).toMatchSnapshot()
})
