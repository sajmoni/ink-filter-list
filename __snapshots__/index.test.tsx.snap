// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Arrow keys and filtering 1`] = `
"[90mType to filter[39m
[34mFirst[39m
[37mSecond[39m
[37mThird[39m
[37mFourth[39m"
`;

exports[`Arrow keys and filtering 2`] = `
"
 [41m[37m ERROR [39m[49m stdin.ref is not a function

 [2mnode_modules/ink/src/components/App.tsx:166:11[22m

 [2m- value[90m (node_modules/ink/src/components/App.tsx:166:11)[22m[39m
 [2m- [90m (node_modules/ink/src/hooks/use-input.ts:128:3)[22m[39m
 [2m-commitHookEffectListMoun[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:14778[22m[39m
                          [90m[2m:26)[22m[39m
 [2m-commitPassiveMountOnFibe[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16609[22m[39m
                          [90m[2m:11)[22m[39m
 [2m-commitPassiveMountEffects_comp[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js[22m[39m
  [2mete[22m                           [90m[2m:16569:9)[22m[39m
 [2m-commitPassiveMountEffects_be[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:1[22m[39m
  [2min[22m                          [90m[2m6556:7)[22m[39m
 [2m-commitPassiveMountEffect[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16544[22m[39m
                          [90m[2m:3)[22m[39m
 [2m-flushPassiveEffectsImpl[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19182:[22m[39m
                         [90m[2m3)[22m[39m
 [2m- flushPassiveEffects[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19127:14)[22m[39m
 [2m- [90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:18912:9)[22m[39m
"
`;

exports[`Arrow keys and filtering 3`] = `
"
 [41m[37m ERROR [39m[49m stdin.ref is not a function

 [2mnode_modules/ink/src/components/App.tsx:166:11[22m

 [2m- value[90m (node_modules/ink/src/components/App.tsx:166:11)[22m[39m
 [2m- [90m (node_modules/ink/src/hooks/use-input.ts:128:3)[22m[39m
 [2m-commitHookEffectListMoun[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:14778[22m[39m
                          [90m[2m:26)[22m[39m
 [2m-commitPassiveMountOnFibe[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16609[22m[39m
                          [90m[2m:11)[22m[39m
 [2m-commitPassiveMountEffects_comp[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js[22m[39m
  [2mete[22m                           [90m[2m:16569:9)[22m[39m
 [2m-commitPassiveMountEffects_be[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:1[22m[39m
  [2min[22m                          [90m[2m6556:7)[22m[39m
 [2m-commitPassiveMountEffect[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16544[22m[39m
                          [90m[2m:3)[22m[39m
 [2m-flushPassiveEffectsImpl[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19182:[22m[39m
                         [90m[2m3)[22m[39m
 [2m- flushPassiveEffects[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19127:14)[22m[39m
 [2m- [90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:18912:9)[22m[39m
"
`;

exports[`Arrow keys and filtering 4`] = `
"
 [41m[37m ERROR [39m[49m stdin.ref is not a function

 [2mnode_modules/ink/src/components/App.tsx:166:11[22m

 [2m- value[90m (node_modules/ink/src/components/App.tsx:166:11)[22m[39m
 [2m- [90m (node_modules/ink/src/hooks/use-input.ts:128:3)[22m[39m
 [2m-commitHookEffectListMoun[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:14778[22m[39m
                          [90m[2m:26)[22m[39m
 [2m-commitPassiveMountOnFibe[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16609[22m[39m
                          [90m[2m:11)[22m[39m
 [2m-commitPassiveMountEffects_comp[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js[22m[39m
  [2mete[22m                           [90m[2m:16569:9)[22m[39m
 [2m-commitPassiveMountEffects_be[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:1[22m[39m
  [2min[22m                          [90m[2m6556:7)[22m[39m
 [2m-commitPassiveMountEffect[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16544[22m[39m
                          [90m[2m:3)[22m[39m
 [2m-flushPassiveEffectsImpl[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19182:[22m[39m
                         [90m[2m3)[22m[39m
 [2m- flushPassiveEffects[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19127:14)[22m[39m
 [2m- [90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:18912:9)[22m[39m
"
`;

exports[`Arrow keys and filtering 5`] = `
"
 [41m[37m ERROR [39m[49m stdin.ref is not a function

 [2mnode_modules/ink/src/components/App.tsx:166:11[22m

 [2m- value[90m (node_modules/ink/src/components/App.tsx:166:11)[22m[39m
 [2m- [90m (node_modules/ink/src/hooks/use-input.ts:128:3)[22m[39m
 [2m-commitHookEffectListMoun[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:14778[22m[39m
                          [90m[2m:26)[22m[39m
 [2m-commitPassiveMountOnFibe[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16609[22m[39m
                          [90m[2m:11)[22m[39m
 [2m-commitPassiveMountEffects_comp[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js[22m[39m
  [2mete[22m                           [90m[2m:16569:9)[22m[39m
 [2m-commitPassiveMountEffects_be[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:1[22m[39m
  [2min[22m                          [90m[2m6556:7)[22m[39m
 [2m-commitPassiveMountEffect[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16544[22m[39m
                          [90m[2m:3)[22m[39m
 [2m-flushPassiveEffectsImpl[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19182:[22m[39m
                         [90m[2m3)[22m[39m
 [2m- flushPassiveEffects[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19127:14)[22m[39m
 [2m- [90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:18912:9)[22m[39m
"
`;

exports[`Arrow keys and filtering 6`] = `
"
 [41m[37m ERROR [39m[49m stdin.ref is not a function

 [2mnode_modules/ink/src/components/App.tsx:166:11[22m

 [2m- value[90m (node_modules/ink/src/components/App.tsx:166:11)[22m[39m
 [2m- [90m (node_modules/ink/src/hooks/use-input.ts:128:3)[22m[39m
 [2m-commitHookEffectListMoun[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:14778[22m[39m
                          [90m[2m:26)[22m[39m
 [2m-commitPassiveMountOnFibe[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16609[22m[39m
                          [90m[2m:11)[22m[39m
 [2m-commitPassiveMountEffects_comp[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js[22m[39m
  [2mete[22m                           [90m[2m:16569:9)[22m[39m
 [2m-commitPassiveMountEffects_be[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:1[22m[39m
  [2min[22m                          [90m[2m6556:7)[22m[39m
 [2m-commitPassiveMountEffect[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:16544[22m[39m
                          [90m[2m:3)[22m[39m
 [2m-flushPassiveEffectsImpl[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19182:[22m[39m
                         [90m[2m3)[22m[39m
 [2m- flushPassiveEffects[90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:19127:14)[22m[39m
 [2m- [90m (node_modules/react-reconciler/cjs/react-reconciler.development.js:18912:9)[22m[39m
"
`;
